modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/QueryIndexRowHandler.java

```diff
@@ -54,31 +49,44 @@ public class QueryIndexRowHandler implements InlineIndexRowHandler {
     /** Index key type settings. */
     private final IndexKeyTypeSettings keyTypeSettings;
 
-    /** H2 Table. */
-    private final GridH2Table h2Table;
-
     /** */
-    public QueryIndexRowHandler(GridH2Table h2table, List<IndexColumn> h2IdxColumns,
-        LinkedHashMap<String, IndexKeyDefinition> keyDefs, List<InlineIndexKeyType> keyTypes, IndexKeyTypeSettings keyTypeSettings) {
-        this.h2IdxColumns = h2IdxColumns;
-        this.keyTypes = keyTypes;
+    public QueryIndexRowHandler(
+        GridQueryRowDescriptor rowDescriptor,
+        LinkedHashMap<String, IndexKeyDefinition> keyDefs,
+        List<InlineIndexKeyType> keyTypes,
+        IndexKeyTypeSettings keyTypeSettings
+    ) {
+        this.keyTypes = Collections.unmodifiableList(keyTypes);
         this.keyDefs = Collections.unmodifiableList(new ArrayList<>(keyDefs.values()));
-        this.h2Table = h2table;
-        cacheDesc = h2table.rowDescriptor();
+        this.rowDescriptor = rowDescriptor;
         this.keyTypeSettings = keyTypeSettings;
+
+        List<Integer> keyColumns = new ArrayList<>(keyDefs.size());
+        List<String> fields = new ArrayList<>(rowDescriptor.type().fields().keySet());
+
+        for (String fieldName : keyDefs.keySet()) {
+            if (QueryUtils.KEY_FIELD_NAME.equals(fieldName))
```

Comment:
Let's move this logic inside `GridQueryRowDescriptor`, it already have similar methods. Also let's move logic of `QueryIndexRowHandler#getKey` to the descriptor too.